pkg/plugins/runtime/gateway/route_table_generator.go

```diff
@@ -110,3 +120,108 @@ func (r *RouteTableGenerator) GenerateHost(ctx xds_context.Context, info *Gatewa
+		var names []string
+		for _, c := range conditions {
+			names = append(names, c.String())
+		}
+		return names
+	}
+
+	configurers := []route.RouteConfigurer{
+		route.RouteActionRetryDefault(protocol),
+	}
+
+	switch protocol {
+	case core_mesh.ProtocolHTTP, core_mesh.ProtocolHTTP2:
+		conf := retry.Spec.GetConf().GetHttp()
+		configurers = append(configurers,
+			route.RouteActionRetryOnStatus(conf.GetRetriableStatusCodes()...),
+			route.RouteActionRetryMethods(methodStrings(conf.GetRetriableMethods())...),
+			route.RouteActionRetryTimeout(conf.GetPerTryTimeout().AsDuration()),
+			route.RouteActionRetryCount(conf.GetNumRetries().GetValue()),
+			route.RouteActionRetryBackoff(
+				conf.GetBackOff().GetBaseInterval().AsDuration(),
+				conf.GetBackOff().GetMaxInterval().AsDuration()),
+		)
+	case core_mesh.ProtocolGRPC:
+		conf := retry.Spec.GetConf().GetGrpc()
+		configurers = append(configurers,
+			route.RouteActionRetryOnConditions(grpcConditionStrings(conf.GetRetryOn())...),
+			route.RouteActionRetryTimeout(conf.GetPerTryTimeout().AsDuration()),
+			route.RouteActionRetryCount(conf.GetNumRetries().GetValue()),
+			route.RouteActionRetryBackoff(
+				conf.GetBackOff().GetBaseInterval().AsDuration(),
+				conf.GetBackOff().GetMaxInterval().AsDuration()),
+		)
+	}
+
+	return configurers
+}
+
+// retryPolicyFor returns the retry policy for the given forwarding
+// action. This is conceptually a bit subtle because a forwarding target
+// can have multiple destinations, each of which is a distinct service.
+// However, there are some relatively obvious rules that we can use to
+// determine policy.
+//
+// 1. If all the destinations are the same service, use that policy.
+// 2. If there are multiple destinations, prefer a wildcard policy.
+// 3. Everything else being equal, older policies are preferred.
+func retryPolicyFor(destinations []route.Destination) *core_mesh.RetryResource {
+	seenNames := map[string]bool{}
+	servicePolicies := map[string][]model.Resource{}
+
+	// Index all the retry policies by service name.
+	for _, d := range destinations {
+		p, ok := d.Policies[core_mesh.RetryType]
+		if !ok {
+			continue
+		}
+
+		if seenNames[p.GetMeta().GetName()] {
+			continue
+		}
+
+		svc := d.Destination[mesh_proto.ServiceTag]
+		servicePolicies[svc] = append(servicePolicies[svc], p)
+		seenNames[p.GetMeta().GetName()] = true
+	}
+
+	var candidates []model.Resource
+
+	// If we are forwarding to multiple services, no one service
+	// would be the most specific match, so we should choose the
+	// wildcard policy. Otherwise, we can just take the oldest of
+	// all the matches, since there's no better way to discriminate.
+	candidates = append(candidates, servicePolicies[mesh_proto.MatchAllTag]...)
+	if len(candidates) == 0 {
+		for _, p := range servicePolicies {
+			candidates = append(candidates, p...)
+		}
+	}
+
+	oldest := match.OldestPolicy(candidates)
```

Comment:
I feel like if `OldestPolicy` can return `nil` then it has to be checked for, otherwise just panic `if len == 0` :shrug:

Reply:
If retry is nil here, we will still end up returning nil to the caller, which handles it. The type case won't panic because it is a checked cast, so it will fail and we will return nil.

Reply:
OK, good point.